### DISCUSSION TITLE
Enhancements for Markdown tables and blockquotes

### DISCUSSION THREAD

**Discussion by razonyang** [hbstack/.github#75] [Announcements]
With the newer table hook and blockquote hook of Hugo `v0.134.0`, now you have a much better control of rendering tables and blockquotes, and generate alerts with blockquotes, rather than using shortcodes.

To use those newest features, you'll need to upgrade Hugo to `v0.134.0` or later, update modules (`hugo mod get -u ...`), and enable `markup.goldmark.parser.attribute.block` as well.

Please refer to the following documents for detailed usage instructions.

- [Markdown tables](https://hbstack.dev/docs/content/tables/)
- [Markdown blockquotes](https://hbstack.dev/docs/content/markdown-syntax/#blockquotes)
- [Markdown blockquote alerts](https://hbstack.dev/modules/content/blockquote-alerts/), a new separate module, as it requires other modules.

Please let us know if there is any issue.

**Comment by TetraTheta** [comment]
FYI, 'Blockquote Render Hook' is added in Hugo v0.132, and 'Table Render Hook' is added in Hugo v0.134.

I really wonder if that added 'Table Render Hook' can create/render 'header-less table' or not...

GFM(GitHub Flavored Markdown) doesn't support it, and Goldmark that Hugo uses also doesn't support it because Goldmark is implementation of CommonMark, which also doesn't support 'header-less table'.

So, the only way to create 'header-less table' is to use HTML directly, and it looks _dirty_ :(

**Reply by razonyang** [reply]
> 'Blockquote Render Hook' is added in Hugo v0.132

Yeah, but our implementation requires AlertTitle and AlertSign, which were introduced in v0.134.0.

> header-less table

Can you provide me with some examples? I believe it can be implemented using attributes.

**Reply by TetraTheta** [reply]
Currently, if I write this markdown in Hugo, it will be converted to this HTML:
```markdown
| Header A | Header B | Header C |
|:---:|:---:|:---:|
| Body A | Body B | Body C |
| Body D | Body E | Body F |
| Body G | Body H | Body I |
```
```html
<table>
  <thead>
    <tr>
      <th style="text-align:center">Header A</th>
      <th style="text-align:center">Header B</th>
      <th style="text-align:center">Header C</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align:center">Body A</td>
      <td style="text-align:center">Body B</td>
      <td style="text-align:center">Body C</td>
    </tr>
    <tr>
      <td style="text-align:center">Body D</td>
      <td style="text-align:center">Body E</td>
      <td style="text-align:center">Body F</td>
    </tr>
    <tr>
      <td style="text-align:center">Body G</td>
      <td style="text-align:center">Body H</td>
      <td style="text-align:center">Body I</td>
    </tr>
  </tbody>
</table>
```
But in `<table>`, both `<thead>` and `<tbody>` are optional. For example, this is valid HTML table without header.
```html
<table>
  <tr>
    <td>Body A</td>
    <td>Body B</td>
    <td>Body C</td>
  </tr>
  <tr>
    <td>Body D</td>
    <td>Body E</td>
    <td>Body F</td>
  </tr>
</table>
```
To create this table in markdown via [kramdown](https://kramdown.gettalong.org/quickref.html#tables), for example, you have to write like this:
```markdown
| Body A | Body B | Body C |
| Body D | Body E | Body F |
```
I think I've saw this syntax for defining text alignment, but I don't remember which parser was that.
```markdown
|:----|:---:|----:|
| Body A | Body B | Body C |
| Body D | Body E | Body F |
| Body G | Body H | Body I |
```

These maybe in area of parser (Goldmark), not render hooks...

**Reply by razonyang** [reply]
This feature is available since github.com/hbstack/hb@v0.15.1. Check https://hbstack.dev/docs/content/tables/#table-without-header out, but you still need to specify the header even if it's empty, to allow the parser to recognize if it's a table.

**Reply by TetraTheta** [reply]
Thank you! 'Specifying empty header' isn't problem at all.
